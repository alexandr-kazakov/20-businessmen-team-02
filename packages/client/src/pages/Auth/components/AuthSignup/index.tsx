import React from 'react'
import { useHistory } from 'react-router-dom'
import { Controller, type FieldValues, useForm } from 'react-hook-form'

import { useAppDispatch } from '../../../../app/redux/hooks'

import { EMAIL_REGEXP, NAME_REGEXP, LOGIN_REGEXP, PHONE_REGEXP, PASSWORD_REGEXP } from '../../../../lib/regexp'
import { signup, setIsSigninView } from '../../redux/authSlice'
import { showSnackBar } from '../../../../components/Snackbar/redux/snackbarSlice'
import { Input } from '../../../../components/UI/Input'
import { Button, ButtonVariant } from '../../../../components/UI/Button'
import { RoutersPaths } from '../../../../components/Routers/types'

import styles from './styles.module.scss'

const INIT_VALUES = {
  email: '',
  login: '',
  first_name: '',
  second_name: '',
  phone: '',
  password: '',
  check_password: '',
}

export const AuthSignUp: React.FC = () => {
  const history = useHistory()
  const dispatch = useAppDispatch()

  const { control, handleSubmit, watch } = useForm({ defaultValues: INIT_VALUES })

  const onSubmit = async (data: FieldValues): Promise<void> => {
    const response = await dispatch(signup(data))

    if (response.error) {
      dispatch(showSnackBar(response.error.message))
    } else {
      history.push(RoutersPaths.main)
    }
  }

  const handlerToggle = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault()
    dispatch(setIsSigninView())
  }

  return (
    <form className={styles.form} onSubmit={handleSubmit(onSubmit)}>
      <span className={styles.title}>Регистрация</span>
      <div className={styles.inputs}>
        <Controller
          name="email"
          control={control}
          rules={{ required: true, pattern: EMAIL_REGEXP }}
          render={({ field, fieldState }) => (
            <Input
              value={field.value}
              onChange={field.onChange}
              placeholder="Почта"
              isValid={!(fieldState.error && 'error')}
            />
          )}
        />

        <Controller
          name="login"
          control={control}
          rules={{ required: true, pattern: LOGIN_REGEXP }}
          render={({ field, fieldState }) => (
            <Input
              value={field.value}
              onChange={field.onChange}
              placeholder="Логин"
              isValid={!(fieldState.error && 'error')}
            />
          )}
        />

        <Controller
          name="first_name"
          control={control}
          rules={{ required: true, pattern: NAME_REGEXP }}
          render={({ field, fieldState }) => (
            <Input
              value={field.value}
              onChange={field.onChange}
              placeholder="Имя"
              isValid={!(fieldState.error && 'error')}
            />
          )}
        />

        <Controller
          name="second_name"
          control={control}
          rules={{ required: true, pattern: NAME_REGEXP }}
          render={({ field, fieldState }) => (
            <Input
              value={field.value}
              onChange={field.onChange}
              placeholder="Фамилия"
              isValid={!(fieldState.error && 'error')}
            />
          )}
        />

        <Controller
          name="phone"
          control={control}
          rules={{ required: true, pattern: PHONE_REGEXP }}
          render={({ field, fieldState }) => (
            <Input
              value={field.value}
              onChange={field.onChange}
              placeholder="Телефон"
              isValid={!(fieldState.error && 'error')}
            />
          )}
        />

        <Controller
          name="password"
          control={control}
          rules={{ required: true, pattern: PASSWORD_REGEXP }}
          render={({ field, fieldState }) => (
            <Input
              type="password"
              value={field.value}
              onChange={field.onChange}
              placeholder="Пароль"
              isValid={!(fieldState.error && 'error')}
            />
          )}
        />

        <Controller
          name="check_password"
          control={control}
          rules={{
            required: true,
            pattern: PASSWORD_REGEXP,
            validate: val => watch('password') === val,
          }}
          render={({ field, fieldState }) => (
            <Input
              type="password"
              value={field.value}
              onChange={field.onChange}
              placeholder="Пароль (ещё раз)"
              isValid={!(fieldState.error && 'error')}
            />
          )}
        />
      </div>
      <div className={styles.buttons}>
        <Button type="submit">Зарегистрироваться</Button>
        <Button onClick={handlerToggle} variant={ButtonVariant.SECONDARY}>
          Войти
        </Button>
      </div>
    </form>
  )
}
