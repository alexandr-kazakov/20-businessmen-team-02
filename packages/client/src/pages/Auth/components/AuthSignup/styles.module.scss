.auth {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100vw;
  height: 100vh;
}

.form {
  display: flex;
  flex-direction: column;
  max-width: 340px;
  width: 100%;
  border: 1px solid #eee;
  border-radius: 12px;
  padding: 30px;
  box-shadow: 0px 0px 6px rgba(0, 0, 0, 0.14);
}

.title {
  text-align: center;
  margin-bottom: 20px;
  font-style: normal;
  font-weight: 500;
  font-size: 20px;
  line-height: 20px;
  color: #fff;
}

.inputs {
  display: flex;
  flex-direction: column;
  gap: 10px;
  margin-bottom: 20px;
}

.buttons {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 10px;
}
