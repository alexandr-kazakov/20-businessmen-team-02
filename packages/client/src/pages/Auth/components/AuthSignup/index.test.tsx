import React from 'react'
import { render } from '@testing-library/react'
import { AuthSignUp } from '@/pages/Auth/components/AuthSignup/index'

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}))

test('should render AuthSignUp component', () => {
  const { asFragment } = render(<AuthSignUp />)

  expect(asFragment()).toMatchSnapshot()
})
