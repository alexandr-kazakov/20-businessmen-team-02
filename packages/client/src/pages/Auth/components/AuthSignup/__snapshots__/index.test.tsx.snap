// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AuthSignUp component 1`] = `
<DocumentFragment>
  <form>
    <span>
      Регистрация
    </span>
    <div>
      <input
        class=""
        placeholder="Почта"
        value=""
      />
      <input
        class=""
        placeholder="Логин"
        value=""
      />
      <input
        class=""
        placeholder="Имя"
        value=""
      />
      <input
        class=""
        placeholder="Фамилия"
        value=""
      />
      <input
        class=""
        placeholder="Телефон"
        value=""
      />
      <input
        class=""
        placeholder="Пароль"
        type="password"
        value=""
      />
      <input
        class=""
        placeholder="Пароль (ещё раз)"
        type="password"
        value=""
      />
    </div>
    <div>
      <button
        class=""
        type="submit"
      >
        Зарегистрироваться
      </button>
      <button
        class="undefined"
      >
        Войти
      </button>
    </div>
  </form>
</DocumentFragment>
`;
